crates/chain/src/canonical_iter.rs

```diff
@@ -0,0 +1,319 @@
+                    hash_map::Entry::Occupied(_) => return None,
+                    hash_map::Entry::Vacant(entry) => entry,
+                };
+                // Any conflicts with a canonical tx can be added to `not_canonical`. Descendants
+                // of `not_canonical` txs can also be added to `not_canonical`.
+                for (_, conflict_txid) in self.tx_graph.direct_conflicts(&tx) {
+                    TxDescendants::new_include_root(
+                        self.tx_graph,
+                        conflict_txid,
+                        |_: usize, txid: Txid| -> Option<()> {
+                            if self.not_canonical.insert(txid) {
+                                Some(())
+                            } else {
+                                None
+                            }
+                        },
+                    )
+                    .run_until_finished()
+                }
+                canonical_entry.insert((tx, this_reason));
+                self.queue.push_back(this_txid);
+                Some(())
+            },
+        )
+        .run_until_finished()
+    }
+}
+
+impl<'g, A: Anchor, C: ChainOracle> Iterator for CanonicalIter<'g, A, C> {
+    type Item = Result<(Txid, Arc<Transaction>, CanonicalReason<A>), C::Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if let Some(txid) = self.queue.pop_front() {
+                let (tx, reason) = self
+                    .canonical
+                    .get(&txid)
+                    .cloned()
+                    .expect("reason must exist");
+                return Some(Ok((txid, tx, reason)));
+            }
+
+            if let Some((txid, tx, anchors)) = self.pending_anchored.next() {
+                if !self.is_visited(txid) {
+                    if let Err(err) = self.scan_anchors(txid, tx, anchors) {
+                        return Some(Err(err));
+                    }
+                }
+                continue;
+            }
+
+            if let Some((txid, tx, last_seen)) = self.pending_last_seen.next() {
+                if !self.is_visited(txid) {
+                    let lsi = LastSeenIn::Mempool(last_seen);
+                    self.mark_canonical(tx, CanonicalReason::from_last_seen(lsi));
+                }
+                continue;
+            }
+
+            if let Some((txid, tx, height)) = self.pending_remaining.pop_front() {
+                if !self.is_visited(txid) {
+                    let lsi = LastSeenIn::Block(height);
+                    self.mark_canonical(tx, CanonicalReason::from_last_seen(lsi));
+                }
+                continue;
+            }
+
+            return None;
+        }
+    }
+}
+
+/// Represents when and where a given transaction is last seen.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, core::hash::Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub enum LastSeenIn {
```

Comment:
Name suggestion to avoid overloading the term last-seen. Also why does it implement serde traits?

```suggestion
pub enum ObservedIn {
```

Reply:
Changed.